src/Chapter3.hs

```diff
@@ -480,6 +517,41 @@ After defining the city, implement the following functions:
    and at least 10 living people inside
 -}
 
+data OptionalCastle = NoCastle | Castle String deriving (Show, Eq)
+data OptionalWall = NoWall | Wall deriving (Show, Eq)
+data RestHouse = Church | Library deriving (Show, Eq)
+data House = MkHouse 
+  { houseLiving :: Int
+  } deriving (Show, Eq)
+
+data MagicalCity = MkCity
+  { cityCastle :: OptionalCastle
+  , cityWall :: OptionalWall
+  , cityRestHouse :: RestHouse
+  , cityHouses :: [House]
+  } deriving (Show, Eq)
+
+buildCastle :: String -> MagicalCity -> MagicalCity
+buildCastle name mc = mc {cityCastle = Castle name} 
+
+buildHouse :: Int -> MagicalCity -> MagicalCity
+buildHouse living mc = mc {cityHouses = newHouseList }
+  where newHouseList = if living > 0 && living <=4 
+        then (MkHouse living) : cityHouses mc 
+        else cityHouses mc
```

Comment:
~Interesting approach. It is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂